[PNUDEV-11] send message to many users

vanya706 opened the pull request:
![image](https://user-images.githubusercontent.com/57075862/107083954-14e7f780-67ff-11eb-80bc-2a112804a254.png)
![image](https://user-images.githubusercontent.com/57075862/107083996-2204e680-67ff-11eb-85ed-4b1b0cf53ddf.png)
![image](https://user-images.githubusercontent.com/57075862/107084052-32b55c80-67ff-11eb-9adf-892ac1ffbd96.png)

vanya706 committed:
[PNUDEV-11] send message to many users

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageServiceImpl.java:
```diff
@@ -1,47 +1,89 @@
+        outboxMessageRepository.save(outboxMessage);
 
-        Message message = Message.builder()
+        if (!notFoundUsernames.isEmpty()) {
```
Throwing exception in successful flow is not very good.
For the design you choose, it's better to create Inbox message with sending error here.

vanya706 committed:
some edits

In MessageServiceImpl and AsyncConfig

vanya706 committed:
edit text

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageServiceImpl.java:
```diff
@@ -1,47 +1,97 @@
+                            .receiverUsername(senderUser.getUsername())
+                            .build(),
+                    senderUser,
+                    userRepository.findByUsername("Gmail Support").orElseThrow(RuntimeException::new)
```
this .`orElseThrow(RuntimeException::new)` can cause problems.

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/config/AsyncConfig.java:
```diff
@@ -24,6 +24,13 @@ public AsyncConfig(ApplicationContext context) {
         this.context = context;
     }
 
+    @Override
```
I believe we can have only logger in this class. No need for any logic, because those cases are handled already.

vanya706 committed:
add logger slf4j

remove redundant logic from AsyncConfig

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageServiceImpl.java:
```diff
@@ -1,47 +1,101 @@
+                        .ifPresentOrElse(receiverUsers::add, () -> notFoundUsernames.add(username))
+        );
+
+        receiverUsers.forEach(receiverUser -> createInboxMessage(messageFormDto, receiverUser, senderUser));
```
Sending should be async, because, in case of the exception when sending message to one of the receivers, we should proceed with seding this message to others anyway.
Obviously, we can catch exception right inside this lambda, but async sending will be better

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/config/AsyncConfig.java:
```diff
@@ -1,55 +1,25 @@
     public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {
         return (throwable, method, objects) -> {
 
-            if (throwable instanceof MessageSendingException){
```
actually, this is needed for handing exceptions from `createInboxMessage `method itself

vanya706 committed:
redesign message sending process

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/exception/handler/ExceptionInterceptor.java:
```diff
@@ -16,7 +16,7 @@
                                                 RedirectAttributes redirectAttributes){
 
-        log.error("Error while submission processing", messageSendingException);
+        log.trace("Error while submission processing", messageSendingException);
```
`log.error` was good

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageSendServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+
+        if (!notFoundUsernames.isEmpty()) {
+
+            User gmail_support = userRepository.findByUsername("Gmail Support")
```
Do you have to have this user in DB?
Cannot it just be some inmemory user?

Igor-Polatajko replied:
naming...

vanya706 committed:
change some logic to logging and renaming field

vanya706 committed:
add ROLE_ADMIN and admin functionality

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageSendServiceImpl.java:
```diff
@@ -0,0 +1,83 @@
+
+            String adminUsername = environment.getProperty("admin.default.username");
+
+            User adminUser = userRepository.findByUsername(adminUsername)
```
No need in such complexity.
This approach solves different task: creating default admin user, for being able to configure software, once you started it for the first time (another approach might be to have a migration for creating this default user, but you won't be able to use env variables to set password and username).

For your problem you need just:
```
private static final User SUPPORT_USER = User.builder().
.....
.build();
```

vanya706 committed:
Revert "add ROLE_ADMIN and admin functionality"

This reverts commit 1912bd87

vanya706 committed:
some edits

add field GMAIL_SUPPORT_ID
add findById into UserService
change parameters in method createInboxMessage
change parameter in method createOutboxMessage
remove field userService from AsyncConfig

Igor-Polatajko commented on src/main/java/com/ivanmostovyi/demo/service/MessageService.java:
```diff
@@ -1,19 +1,23 @@
 public interface MessageService {
 
-    void create(MessageFormDto messageFormDto, User senderUser);
+    Long GMAIL_SUPPORT_ID = 1L;
```
Add migration to create this user.
Disable login for this user.

Igor-Polatajko replied:
Also, use username here, not id, because we don't want to specify id in the migration

vanya706 committed:
add new migration "V2_MyGmailSupportUser" and implements it in application

Igor-Polatajko commented on src/main/resources/db/migration/V2__MyGmailSupportUser.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO "user" (username, password, role, locked, active)
```
Won't `active=false`,` locked=true` do the same as your check in the service?
But it's better to keep that check in the service anyway.
